autoencoder.py

```diff
@@ -22,51 +37,132 @@ def __init__(self, intermediate_dim, code_dim, original_dim, num_hidden_layers,
-          layer = layer + 1
-          activation = self.decoder_hidden_layers[layer](activation)
-      return self.decoder_output_layer(activation)
+        """Encoder"""
+
+        activation = self.encoder_hidden_layers[0](input_features)
+        if self.num_hidden_layers > 1:
+            for layer in range(self.num_hidden_layers - 1):
+                layer = layer + 1
+                activation = \
+                    self.encoder_hidden_layers[layer](activation)
+        activation = self.encoder_output_layer(activation)
+        activation = self.decoder_hidden_layers[0](activation)
+        if self.num_hidden_layers > 1:
+            for layer in range(self.num_hidden_layers - 1):
+                layer = layer + 1
+                activation = \
+                    self.decoder_hidden_layers[layer](activation)
+        return self.decoder_output_layer(activation)
+
+    def opt(self):
+        return self.optimizer
+
 
 def loss(reconstructed, original):
-    return tf.reduce_mean(tf.square(tf.subtract(reconstructed, original)))
+    return tf.reduce_mean(tf.square(tf.subtract(reconstructed,
+                          original)))
+
+
+def train(
+    loss,
+    model,
+    opt,
+    original,
+    ):
 
-def train(loss, model, opt, original):
     with tf.GradientTape() as tape:
         reconstructed = model(original)
         reconstruction_error = loss(reconstructed, original)
-    gradients = tape.gradient(reconstruction_error, model.trainable_variables)
+    gradients = tape.gradient(reconstruction_error,
+                              model.trainable_variables)
     gradient_variables = zip(gradients, model.trainable_variables)
     opt.apply_gradients(gradient_variables)
+
+    train_loss(reconstruction_error)
+
     return reconstruction_error
 
-def train_loop(model, loss, dataset, epochs):
-  opt = tf.optimizers.Adam(learning_rate=1e-3)
-  for epoch in range(epochs):
-      epoch_loss = 0
-      for batch_features in dataset:
-          loss_values = train(loss, model, opt, batch_features)
-          epoch_loss += loss_values
-      model.loss.append(tf.reduce_mean(epoch_loss))
-      print('Epoch {}/{}. Loss: {}'.format(epoch + 1, epochs, tf.reduce_mean(epoch_loss)))
\ No newline at end of file
+
+def train_loop(
+    model,
+    loss,
+    dataset,
+    epochs,
+    ):
+
+    opt = model.opt()
+    for epoch in range(epochs):
+        epoch_loss = 0
+        for batch_features in dataset:
+            loss_values = train(loss, model, opt, batch_features)
+            with train_summary_writer.as_default():
+                tf.summary.scalar('loss', train_loss.result(),
+                                  step=epoch)
+            epoch_loss += loss_values
+        model.loss.append(tf.reduce_mean(epoch_loss))
+        print('Epoch {}/{}. Loss: {}'.format(epoch + 1, epochs,
+              tf.reduce_mean(epoch_loss)))
+        train_loss.reset_states()
+
+
+def log_images(test_features, model):
```

Comment:
Let's move this to a different module, say, `utils.py` @ChadElRicardo

Reply:
Hi @AFAgarap, is the **utils.py** will be containing both `train` & `train_step` function?

Reply:
Yes. They can be moved there, so we can reuse them.